### PR TITLE
Commit messages

### PR DESCRIPTION
**Pull request by manfromjupyter**

### CODE REVIEW

**Comment by manfromjupyter** [comment]
Note to self - need to also include in commit message if it was a change request and who approved it

**Comment by manfromjupyter** [comment]
@mcrutch Assigning you but might need your opinion first on how to handle Change Requests in terms of the view within the /notebooks/{ID}/revisions page. So we had feedback on the site to record who accepted change requests and display it with the revision history to better understand the story. You can veto that, of course, but I just wanted to see if you like it as is or if we should add another column that's just a check mark if or not it was due to a change request where on hover would reveal who approved it? My current way of populating that column:

"{CHANGE LOG} - via change request (accepted by {USER})."
example: Gave it a second head so it can read logs twice as fast - via change request (accepted by John Smith).

Can go to notebook ID: 464's revisions to see how it looks.

In terms of what this branch needs to get double checked on in terms of testing is:
1. New uploads auto-generate a commit of "Notebook created"
2. New version updates require you to enter a new Change Log/commit message.
3. Change Requests do the same but have the added step of needing to be approved.
a. (UNTESTED) The initial change request email needs to be checked that it contains the Change Log in it.
4. Commit messages can be changed within the /revisions page.

I'm doing this in kind of a rush before I leave so if I forgot to delete a comment or something, forgive me hahahah.

**Comment by mcrutch** [comment]
Please merge with latest main (that will let me test email but this branch started so long ago it doesn't have some necessary code)
"Accepted by (USER)"  Since you are storing this in a plain-text field you can't link to the user ui but at the same time, this can be ambigous using the display name for the user, unless it also links to their page.  I feel like, if we are going to have that information there, it needs to be "right from the get-go" and right now, it is ambiguous. and since the commit message can be edited free-text...don't know that I want it at all as part of the text.

The UI for updating commit messages is not suitable (see screenshot)
image

**Comment by mcrutch** [comment]
Also.....
image
HTML/Javascript injection

**Comment by mcrutch** [comment]
Confirmed email message has the appropriate information:
image

**Comment by manfromjupyter** [comment]
Finally ready for review after numerous testing and perfectionisting

## Closes
Closes: #180, #376, #384, and others unreported

##  Commit Messages Piece (#180)
- Uploading a new notebook created a automated commit message that you created the notebook, any future uploads of new version and update via change request allows the user to add a Change Log which will be that revision's commit message. 
- Visiting Revision History will allow you these commit messages per revision and added ability to edit the commit messages.
- Added it so updates via Change Request now permanently apply a reference at the end of the revision history of the original change request that that change log originated from and in that is the user that submitted the change request and who accepted the change request (which fixes another issue I don't have filed here).

## Additional Change Request refactor and updates
- Removed "Decline" option for Change Request if you're the one that Created the Change Request (#384). No more double red buttons side by side on those pages.
- Refactored the Change Request emails. They are all consistent with one another and reveal more information for the user than they did before such as what request a user was cancelling, what the metadata is associated with all of these, and if accepted, canceled, or declined, includes the information of the change request the owner is commenting on so the submitter and other owners understand what exactly they're talking about.
- Sorted the regular Change Requests page (not the change_requests/all) tables by date so most recent stuff is always at the top. Making a change would cause the most recent thing suddenly go down several listings and get lost if you had multiple change requests for the same notebook, which many users probably do; keeping ones needing action at the top just makes it so disorganized.
- Changed date of when submitter submitted change request from updated_at to created_at so it's more accurate. When owner leaves comment, theirs is the updated_at date.
- Cleaned up the Change Request/ID landing pages after they've been approved, declined, or canceled and added links back to the archived change requests so users can visit to see in a more reader-friendly view the timeline and the comments from each party for the given change request (#376)


## Where are the fields
upload change request - new
**NOTE** - The "Comment" has been switch back to "Comments". Ignore that.
upload new version - new


## Revision Page
revision page for commit messages
revison page on edit


## Changes requests show page for when you are both submitted and owner
changes requests show page for when you are both submitted and owner


## Changes requests show page for when you are owner but not submitter
changes requests show page for when you are owner but not submitter


## Emails
new change request email
change request declined email
change request cancelled email
change request accepted email
**NOTE** - spacing between Owner comments and Change Request section below has been fixed to match that of the Declined email


## Archived change requests
change request links to archived requests
change request declined afterwards
change request accepted afterwards and if one of the owners is deleted
change request cancelled afterwards


### If reviewer/one of the owners is deleted
change request accepted afterwards and if one of the owners is deleted

File: app/controllers/notebooks_controller.rb [hunk 1]
```diff
@@ -136,12 +139,25 @@ def update
     @old_content = @notebook.content
     @tags = parse_tags
     populate_notebook
-
-    # Save the content and db record.
-    if save_update
+    errors = " "
+    summary = params[:summary].strip
+    if summary.length > 500
+      errors += "Change log was too long. Only accepts 500 characters and you tricked the form to submit one that was #{summary.length} characters."
```
Comment (mcrutch) [inline app/controllers/notebooks_controller.rb:145]: Don't say "you tricked the form".  Just leave it as "500 characters and you submitted on that was ...."

File: app/controllers/revisions_controller.rb [hunk 2]
```diff
@@ -57,6 +57,32 @@ def latest_diff
     end
   end
 
+  # PATCH /notebooks/:notebook_id/revisions/:commit_id/edit_summary
+  def edit_summary
+    errors = ""
+    revision_summary = params[:summary].strip
+    if revision_summary.length > 500
+      errors += "Revision summary was too long. Only accepts 500 characters and you tricked the form to submit one that was #{revision_summary.length} characters."
```
Comment (mcrutch) [inline app/controllers/revisions_controller.rb:65]: Again with the tricked the form thing

File: app/controllers/notebooks_controller.rb (hunk 1 above)
Comment (manfromjupyter) [reply app/controllers/notebooks_controller.rb:145]: Gotcha gotcha. Makes sense. More professional

Comment (mcrutch) [reply app/controllers/notebooks_controller.rb:145]: Not just that, it automatically assigns blame to the user when at some point a Javascript glitch could occur that allows it through anyway.

Comment (manfromjupyter) [reply app/controllers/notebooks_controller.rb:145]: I just assume people that get crazy bugs are just doing crazy things, probably a bunch of hackers. Makes sense though. Will make the changes

**Comment by manfromjupyter** [comment]
@mcrutch, I removed the trickery. Should be good to go

**Comment by manfromjupyter** [comment]
Did a lot of merges that I want a sanity check on. Going to test this one more time before merging

**Comment by manfromjupyter** [comment]
@mcrutch After many merge conflicts and fixing my failed merge fixes this is now ready.

**Comment by manfromjupyter** [comment]
@myself, don't let users edit metadata revisions as they're the exact same as the one below them and cause weird issues. Just disable and gray out those cells. Make sure it's pretty

**Comment by manfromjupyter** [comment]
@mcrutch do any of these look good (for default theme)? If we just want to disable that one field instead of all of this, let me know, but figured since the other fields are empty (always) might as well.

revisions metadata solution 2 - lighter
revisions metadata solution 1
revisions metadata solution 3

**Comment by mcrutch** [comment]
Keep the columns, so 1 or 2. I am leaning for the second option, just sanity check the other themes.

**Comment by manfromjupyter** [comment]
@mcrutch If you like the looks of it, it's good to go

revisions metadata final dark theme
revisions metadata final default
revisions metadata final ultra dark theme

**Pull request merged**